### PR TITLE
Add Save events for later feature

### LINKED ISSUE #87: Save events
As a user, I want to save events to view them easily later without necessarily joining them. 

Requirement: on the event page, a button that sends the event to the Save for later page. This page is accessible from the profile page.

### PR DESCRIPTION
**Pull request by Acheline**
# Summary

This PR adds end-to-end support for saving/un-saving events per user, including repository APIs, Firestore + local implementations, ViewModel state, UI controls, and test coverage. It also wires the feature into the bottom sheet so users can view and open their saved events. Closes #87 .

# Changes
### Repository API

**Added methods to EventRepository:**

- getSavedEventIds(userId: String): Set<String>

- getSavedEvents(userId: String): List<Event>

- saveEventForUser(userId: String, eventId: String): Boolean

- unsaveEventForUser(userId: String, eventId: String): Boolean

### Firestore implementation

**EventRepositoryFirestore:**

- Saves per-user docs under users/{userId}/savedEvents/{eventId} with savedAt timestamp.

- getSavedEventIds reads subcollection IDs.

- getSavedEvents batches whereIn(FieldPath.documentId()) with chunking (limit 10), merges, and sorts by date.

- saveEventForUser/unsaveEventForUser are idempotent; return true/false on success/failure.

### Local implementation

**LocalEventRepository:**

- In-memory per-user LinkedHashSet of saved IDs.

- Implements the new API (idempotent adds/removes) for offline/dev use.

### UI & ViewModel

**Event detail sheet**

- Shows Save for later (primary button) when not saved.

- Shows Unsave (outlined, error color) when saved.

- Logic extracted via resolveSaveButtonUi(isSaved); covered by unit tests.

**Bottom sheet**

- New Saved Events tab (default selection).

- Shared EventsSection used for both Saved + Joined lists with “Show more” affordance.

**MapScreenViewModel**

Tracks:

- _savedEventIds for O(1) checks, exposed via isEventSaved(event).

- _savedEvents list for the Saved tab.

- Loads saved IDs + events on init when user is signed in.

- Implements saveEventForLater() and unsaveEventForLater() with error messages for common edge cases (not signed in, already saved, etc.).

- Keeps _events / _allEvents intact; saved state is independent of participation.

- onTabEventClicked(event) focuses the map pin and opens details from either Saved or Joined tabs.

### UX details

- “Save for later” visible in FULL sheet.

- Save → refreshes saved list; Unsave → removes and refreshes.

- Error messages surfaced through existing ViewModel _errorMessage.

### Data Model

**Firestore:**

- users/{userId}/savedEvents/{eventId} with field savedAt: serverTimestamp().

**No changes to Event schema.**

**Limit constants:**

- Query LIMIT = 10 reused for tag and whereIn chunking.

### Modified Files (high level)

**Core**
```
app/src/main/java/com/swent/mapin/model/event/EventRepository.kt (+ API)

app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt

app/src/main/java/com/swent/mapin/model/event/LocalEventRepository.kt
```
**UI**
```
app/src/main/java/com/swent/mapin/ui/map/EventDetailSheet.kt

app/src/main/java/com/swent/mapin/ui/map/BottomSheetContent.kt

app/src/main/java/com/swent/mapin/ui/map/MapScreen.kt

app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt
```
**Tests**
```
app/src/test/java/com/swent/mapin/model/EventRepositoryFirestoreTest.kt

app/src/test/java/com/swent/mapin/model/event/LocalEventRepositoryTest.kt

app/src/test/java/com/swent/mapin/ui/map/EventDetailSheetLogicTest.kt

app/src/test/java/com/swent/mapin/ui/map/MapScreenViewModelTest.kt
```
### Tests

**Firestore repository**

- Maps saved IDs; handles failures → empty set.

- saveEventForUser/unsaveEventForUser success/failure paths.

- getSavedEvents with chunking over the 10-ID whereIn limit, merges + sorts by date.

**Local repository**

- Add/Idempotency/Remove; per-user isolation; missing event ID returns false; preserves insertion order.

**UI logic**

- resolveSaveButtonUi for saved vs. not saved.

**ViewModel**

- Save/Unsave success → updates _savedEventIds, refreshes _savedEvents, clears errors.

- Error scenarios: not signed in, already saved, unsave when not saved.

- “Saved Events” tab default; click from Saved tab centers pin via onTabEventClicked.

- Overall coverage: adds robust unit tests across repository, UI logic, and ViewModel, including Firestore edge cases and chunking behavior.

# Demo
Screenshot 2025-10-27 212551

Screenshot 2025-10-27 212635

## Summary by CodeRabbit

* **New Features**
  * Users can save and unsave events; saved events are persisted per user and shown in a Saved Events tab.

* **Updates**
  * Default bottom sheet tab changed from "Recent Activities" to "Saved Events."
  * Event detail sheet shows current save state and toggles between "Save for Later" and "Unsave."

* **Tests**
  * Expanded test coverage for save/unsave flows, tab switching, and saved-events retrieval and UI.

### CODE REVIEW

File: app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt [hunk 1]
```diff
@@ -617,7 +684,11 @@ class MapScreenViewModel(
         eventRepository.editEvent(event.uid, updatedEvent)
         _selectedEvent = updatedEvent
         _events = _events.map { if (it.uid == event.uid) updatedEvent else it }
+        _allEvents = _allEvents.map { if (it.uid == event.uid) updatedEvent else it }
         loadJoinedEvents()
+        for (event in _joinedEvents) {
+          Log.d("test", "Joined event after join: ${event.uid}")
+        }
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt]: - Error: Leftover debug Log.d calls with tag "test" remain in production code. These logs appear to be temporary and create noise in runtime logs.
- Suggestion: Remove these debug statements or replace with a clear, intentional tag and appropriate log level. Avoid ad-hoc logging inside loops that execute on every join/unregister.

File: app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt [hunk 2]
```diff
@@ -182,10 +194,41 @@ class MapScreenViewModel(
...
+            _joinedEvents = emptyList()
+          } else {
+            // Signed in → (re)load user-scoped data
+            loadSavedEvents()
+            loadSavedEventIds()
+            loadJoinedEvents()
+            loadParticipantEvents()
+          }
+        }
+    auth.addAuthStateListener(authListener!!)
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt]: - Nitpick: The auth listener is registered during init and removed in onCleared, which is correct, but the guard if (authListener != null) return relies on instance identity only.
- Suggestion: Add small comment documenting the reason for the guard and that the ViewModel lifecycle controls listener removal, to avoid accidental double-registration in future refactors.

File: app/src/test/java/com/swent/mapin/ui/map/EventDetailSheetLogicTest.kt
Comment (naderbahri) [inline app/src/test/java/com/swent/mapin/ui/map/EventDetailSheetLogicTest.kt]: General testing and edge cases

- Question: Are there tests that cover concurrent update scenarios where events are reloaded from repo while a save/unsave request is in-flight
- Suggestion: Add tests that simulate repository throwing exceptions during save/unsave and verify ViewModel handles rollback or proper errorMessage for the user. Add tests for getSavedEvents when some event IDs returned by saved subcollection do not correspond to any event in the events collection, and for large saved lists that require chunking (already present for chunking, but add a variant where some ids missing).

File: app/src/main/java/com/swent/mapin/ui/map/EventDetailSheet.kt [hunk 3]
```diff
@@ -410,15 +416,38 @@ private fun FullEventContent(
...
+
+@VisibleForTesting
+internal fun resolveSaveButtonUi(isSaved: Boolean): SaveButtonUi {
+  return if (isSaved) {
+    SaveButtonUi(showSaveButton = false, label = "Unsave")
+  } else {
+    SaveButtonUi(showSaveButton = true, label = "Save for later")
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (Acheline) [reply app/src/main/java/com/swent/mapin/ui/map/EventDetailSheet.kt]: showPrimarySaveAction and showSaveButton is the same. I don't see the confusion here.

File: app/src/test/java/com/swent/mapin/ui/map/EventDetailSheetLogicTest.kt
Comment (Acheline) [reply app/src/test/java/com/swent/mapin/ui/map/EventDetailSheetLogicTest.kt]: yes there are.
